pkg/suggestion/v1alpha3/goptuna/converter.go

```diff
@@ -0,0 +1,282 @@
+			high, err := strconv.Atoi(p.GetFeasibleSpace().GetMax())
+			if err != nil {
+				return nil, err
+			}
+			low, err := strconv.Atoi(p.GetFeasibleSpace().GetMin())
+			if err != nil {
+				return nil, err
+			}
+			stepstr := p.GetFeasibleSpace().GetStep()
+			if stepstr == "" {
+				searchSpace[p.Name] = goptuna.IntUniformDistribution{
+					High: high,
+					Low:  low,
+				}
+			} else {
+				step, err := strconv.Atoi(stepstr)
+				if err != nil {
+					return nil, err
+				}
+				searchSpace[p.Name] = goptuna.StepIntUniformDistribution{
+					High: high,
+					Low:  low,
+					Step: step,
+				}
+			}
+		} else if p.ParameterType == api_v1_alpha3.ParameterType_CATEGORICAL {
+			choices := p.GetFeasibleSpace().GetList()
+			searchSpace[p.Name] = goptuna.CategoricalDistribution{
+				Choices: choices,
+			}
+		} else if p.ParameterType == api_v1_alpha3.ParameterType_DISCRETE {
+			// Use categorical distribution instead of goptuna.DiscreteUniformDistribution
+			// because goptuna.UniformDistributions needs to declare the parameter space
+			// with minimum value, maximum value and interval.
+			choices := p.GetFeasibleSpace().GetList()
+			searchSpace[p.Name] = goptuna.CategoricalDistribution{
+				Choices: choices,
+			}
+		} else {
+			return nil, errors.New("unsupported parameter type")
+		}
+	}
+	return searchSpace, nil
+}
+
+func toGoptunaState(condition api_v1_alpha3.TrialStatus_TrialConditionType) (goptuna.TrialState, error) {
+	if condition == api_v1_alpha3.TrialStatus_CREATED {
+		return goptuna.TrialStateWaiting, nil
+	} else if condition == api_v1_alpha3.TrialStatus_RUNNING {
+		return goptuna.TrialStateRunning, nil
+	} else if condition == api_v1_alpha3.TrialStatus_SUCCEEDED {
+		return goptuna.TrialStateComplete, nil
+	} else if condition == api_v1_alpha3.TrialStatus_KILLED {
+		return goptuna.TrialStateFail, nil
+	} else if condition == api_v1_alpha3.TrialStatus_FAILED {
+		return goptuna.TrialStateFail, nil
+	}
+	return goptuna.TrialStateFail, errors.New("unexpected trial condition")
+}
+
+func toGoptunaTrials(
+	ktrials []*api_v1_alpha3.Trial,
+	study *goptuna.Study,
+	searchSpace map[string]interface{},
+) ([]goptuna.FrozenTrial, error) {
+	gtrials := make([]goptuna.FrozenTrial, 0, len(ktrials))
+	for i, kt := range ktrials {
+		datetimeStart, err := time.Parse(time.RFC3339Nano, kt.GetStatus().GetStartTime())
+		if err != nil {
+			return nil, err
+		}
+		datetimeComplete, err := time.Parse(time.RFC3339Nano, kt.GetStatus().GetCompletionTime())
+		if err != nil {
+			return nil, err
+		}
+		state, err := toGoptunaState(kt.GetStatus().GetCondition())
+		if err != nil {
+			return nil, err
+		}
+
+		metrics := kt.GetStatus().GetObservation().GetMetrics()
+		intermediateValues := make(map[int]float64, len(metrics))
+		var finalValue float64
+		for i, m := range metrics {
+			v, err := strconv.ParseFloat(m.GetValue(), 64)
+			if err != nil {
+				return nil, err
+			}
+			intermediateValues[i] = v
```

Comment:
Unfortunately, you can't get `intermediateValues` right now. Currently, Trial `Observation` is recorded only once Trial is succeeded. Is it fine for the algorithm?

Reply:
Yeah, no problem. `intermediateValues` are just used by early stopping algorithms (Goptuna currently supports Median Stopping Rule and Asynchronous Successive Halving Algorithm).

To simplify the code, I set `nil` in `intermediateValues` at:

https://github.com/kubeflow/katib/pull/1131/commits/6c8e1e830871a316ebe22a4d2bd9ec22a8a918d9

Thank you for your comment!